### PR TITLE
Implement support for Gogs

### PR DESCRIPTION
**Pull request by jiripetrlik**

### CODE REVIEW

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.http.entity.ContentType;
+import org.kie.cloud.git.AbstractGitProvider;
+import org.apache.http.client.fluent.Request;
+import org.kie.cloud.git.constants.GitConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GogsGitProvider extends AbstractGitProvider {
+
+    private String url;
```
Comment (jhrcek) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: Can we make these final, please?

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        createRepository(repositoryName);
+        pushToGitRepository(getRepositoryUrl(repositoryName), repositoryPath,
+                user, password);
+
+        return repositoryName;
+    }
+
+    @Override public void deleteGitRepository(String repositoryName) {
+        try {
+            final int statusCode = Request.Delete(url + URL_API_SUFFIX + URL_REPOSITORY_SUFFIX + "/" + user + "/" +
```
Comment (jhrcek) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: Building URLs using string concatenation can be error prone (e.g. what if some piece ends/doesn't end with '/'?). Unfortunately I'm not aware of java ecosystem that solves this in a nicer way :P

File: pom.xml [hunk 3]
```diff
@@ -160,6 +177,12 @@
         <artifactId>maven-verifier</artifactId>
         <version>${version.org.apache.maven.shared}</version>
       </dependency>
+
+      <dependency>
+        <groupId>org.eclipse.jgit</groupId>
+        <artifactId>org.eclipse.jgit</artifactId>
+        <version>4.11.0.201803080745-r</version>
```
Comment (sutaakar) [inline pom.xml]: Just a nitpick, please put the version into the properties so they all are in one place.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java [hunk 4]
```diff
@@ -98,10 +98,6 @@ public void initializeDeployment() {
     public void cleanEnvironment() {
         if (deploymentScenario != null) {
             deploymentScenario.undeploy();
-
-            if (gitProvider != null) {
-                gitProvider.deleteGitRepository(deploymentScenario.getNamespace());
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java]: Why did you delete this code? Does it cause any issue with Gogs?

Comment (jiripetrlik) [reply test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java]: Yes, it seems that delete method is called twice. At least in some cases. For example in "KieServerS2iJbpmIntegrationTest". It is called here:
- https://github.com/kiegroup/kie-cloud-tests/blob/master/test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/s2i/KieServerS2iJbpmIntegrationTest.java#L103
- https://github.com/kiegroup/kie-cloud-tests/blob/master/test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java#L103

File: pom.xml (hunk 3 above)
Comment (jiripetrlik) [reply pom.xml]: Good catch, will improve it.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java (hunk 4 above)
Comment (sutaakar) [reply test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java]: Ok, but why did you delete it here?

Comment (jiripetrlik) [reply test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/AbstractCloudIntegrationTest.java]: It's also problem that in this place it is trying to delete wrong git project. Will try to get to know more and decide better. :-)

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    private String user;
+    private String password;
+
+    private static final String URL_API_SUFFIX = "/api/v1";
+    private static final String URL_CREATE_REPOSITORY_SUFFIX = "/user/repos";
+    private static final String URL_REPOSITORY_SUFFIX = "/repos";
+
+    private static final Logger logger = LoggerFactory.getLogger(GogsGitProvider.class);
+
+    @Override public String createGitRepositoryWithPrefix(String repositoryPrefixName, String repositoryPath) {
```
Comment (jakubschwan) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: Just a nitpick, from my point of view for reading purpose is better to have Override annotation on the separate line.

File: README.md [hunk 6]
```diff
@@ -49,6 +49,14 @@ Can be found in framework-git, class org.kie.cloud.git.constants.GitConstants
 | github.username     |        | Username for logging into GitHub      |
 | github.password     |        | Password for logging into GitHub      |
 
+#### Gogs
+
+| \<specific-params\> | Value  |  Meaning                              |
+| ------------------- | ------ | ------------------------------------- |
+| git.provider        | Gogs   |                                       |
+| gogs.username       |        | Username for logging into Gogs        |
+| gogs.password       |        | Password for logging into GitHub      |
```
Comment (winklerm) [inline README.md]: Github or Gogs?

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (winklerm) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java]: 2018?

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import org.kie.cloud.git.github.GitHubGitProviderFactory;
+import org.kie.cloud.git.gitlab.GitLabGitProviderFactory;
+import org.kie.cloud.git.gogs.GogsGitProviderFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GitProviderService {
+    private final Map<String, GitProviderFactory> providerFactories;
+
+    private static final Logger logger = LoggerFactory.getLogger(GitProviderService.class);
```
Comment (winklerm) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java]: I would probably suggest LOGGER, since it is a private static final field. But I am not aware of code conventions in this particular repository.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    private final String url;
+    private final String user;
+    private final String password;
+
+    private static final String URL_API_SUFFIX = "/api/v1/";
+    private static final String URL_CREATE_REPOSITORY_SUFFIX = "user/repos";
+    private static final String URL_REPOSITORY_SUFFIX = "repos/";
+    private static final String URL_GIT_SUFFIX = ".git";
+
+    private static final Logger logger = LoggerFactory.getLogger(GogsGitProvider.class);
```
Comment (winklerm) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: The same as before, LOGGER?

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/constants/GitConstants.java [hunk 10]
```diff
@@ -52,6 +60,27 @@ public static String getGitHubPassword() {
...
+
+    public static String getGogsUser() {
+        return System.getProperty(GOGS_USER);
+    }
+
+    public static String getGogsPassword() {
+        return System.getProperty(GOGS_PASSWORD);
+    }
+
+    public static String readMandatoryParameter(String value, String parameterName) {
```
Comment (jhrcek) [inline framework-cloud/framework-git/src/main/java/org/kie/cloud/git/constants/GitConstants.java]: Wouldn't it be better to make this implementation detail of each of the methods? I don't like that the API doesn't enforce the "mandatoriness". As it is you can misuse it by calling `GitConstants.getGogsPassword`and it would fail if not seat. How about this instead?
```
public static String getGogsPassword() {
       return readMandatoryParameter(GOGS_PASSWORD);
}

private static String readMandatoryParameter(String parameterName) {
        Strinv value = System.getProperty(parameterName)
        if (value == null) {
            logger.error("Parameter {} must be specified", parameterName);
            throw new RuntimeException("Parameter " + parameterName + " must be specified");
        }
        return value;
 }
```

File: README.md (hunk 6 above)
Comment (jiripetrlik) [reply README.md]: Good catch. Fixed.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java (hunk 7 above)
Comment (jiripetrlik) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java]: Thanks. :-) Fixed.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/constants/GitConstants.java (hunk 10 above)
Comment (jiripetrlik) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/constants/GitConstants.java]: @jhrcek Yes, you are right. But it is problem of all parameters in Kie cloud tests. I think we can polish it in another PR.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java (hunk 9 above)
Comment (jiripetrlik) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: We use logger with small letters everywhere in this test suite. I will keep it this way.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java (hunk 8 above)
Comment (jiripetrlik) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java]: We use logger with small letters everywhere in this test suite. I will keep it this way.

**Comment by jiripetrlik** [comment]
@winklerm Thank you for valuable comments. Can you please approve now?

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java (hunk 8 above)
Comment (winklerm) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/GitProviderService.java]: Sure, np.

File: framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java (hunk 9 above)
Comment (winklerm) [reply framework-cloud/framework-git/src/main/java/org/kie/cloud/git/gogs/GogsGitProvider.java]: Np!

**Comment by sutaakar** [comment]
@jiripetrlik The PR doesn't delete projects from GitLab repo, do you have any idea what can be wrong?

**Comment by sutaakar** [comment]
retest this please
